### DISCUSSION TITLE
Is the 'bIsNativeBranchingPoint' necessary in AlsAnimNotifyXXX?

### DISCUSSION THREAD

**Discussion by watsonsong** [Sixze/ALS-Refactored#449] [Q&A]
I found almost every AlsAnimNotifyXXX set the `bIsNativeBranchingPoint = true`, including SetLocomotionAction, EarlyBlendOut, SetRootMotionScale, etc.
But the branching point should place only 1 in a single frame. So it could be better to use the bIsNativeBranchingPoint as little as possible. And the branch point is used to keep the order between AnimNotifies, so I think maybe all of them is not care about this.

**Comment by onx2** [comment]
I'm also interested in hearing about this. I've had to use branching point for things that are gameplay critical like triggering a notify for a gameplay ability within a montage but have heard for typical use cases it shouldn't be used?

**Comment by watsonsong** [comment]
The first point is, two branch point can not put in the same time. For example, if you have two NotifyState both start at the beginning, unreal will report warning and tell you only one of them will be trigger.
As far as I know, the branch point used to keep order of other notifies. The notify before and after the point will keep in order, even the notify trigger in one frame.

**Comment by Sixze** [comment]
I've done a little research, and it seems that branching point should be used in notifies that may affect the playback of the current montage. Branching point notifies are executed as early as possible, immediately during the `FAnimMontageInstance::Advance()` function, so they can affect its further playback, while normal notifies are placed in a queue to be processed at the end of the `UAnimInstance` tick.

- `UAlsAnimNotifyState_SetLocomotionAction` is a branching point since it calls `AlsCharacter::OnLocomotionActionChanged()`, which could possibly affect the playback of the current montage, but I'm not 100% sure.
- `UAlsAnimNotifyState_EarlyBlendOut` is a branching point since it could interrupt the montage.
- `UAlsAnimNotifyState_SetRootMotionScale` is not a branching point because changing the root motion scale will only have an effect on the next tick of the character motion component.
- `UAlsAnimNotify_SetGroundedEntryMode` is not a branching point because it changes one variable on the animation instance without side effects.
